### PR TITLE
Updated javadoc and doc files on Scalars

### PR DESCRIPTION
**Pull request by bbakerman**
see #927 

Updated the JavaDoc and documentation on writing custom scalars

### CODE REVIEW

File: docs/scalars.rst [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        });
+
+
+Coercing values
+---------------
+
+The real work in any custom scalar implementation is the ``graphql.schema.Coercing`` implementation.  This is responsible for 3 functions
+
+* ``parseValue`` - takes a variable input object and converts into the Java runtime representation
+* ``parseLiteral`` - takes an AST literal ``graphql.language.Value` as input and converts into the Java runtime representation
```
Comment (andimarek) [inline docs/scalars.rst]: this only happens when the scalar value is defined as a literal as part of the query (as opposed to using a variable)

File: docs/scalars.rst [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            // ps.  I am not trying to replicate RFC-3696 clearly
+            return Pattern.matches("[A-Za-z0-9]@[.*]", possibleEmailValue);
+        }
+
+        private static Object serializeEmail(Object dataFetcherResult) {
+            String possibleEmailValue = String.valueOf(dataFetcherResult);
+            if (looksLikeAnEmailAddress(possibleEmailValue)) {
+                return possibleEmailValue;
+            } else {
+                throw new CoercingSerializeException("Unable to serialise " + possibleEmailValue + " as an email address");
```
Comment (andimarek) [inline docs/scalars.rst]: typo in seralise

File: docs/scalars.rst [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+* The ``serialise`` MUST ONLY allow ``graphql.schema.CoercingSerializeException`` to be thrown from it.  This indicates that the
+value cannot be serialised into an appropriate form.  You must not allow other runtime exceptions to escape this method to get
+the normal graphql behaviour for validation.  You MUST return a non null value
+
+
+* The ``parseValue`` MUST ONLY allow ``graphql.schema.CoercingParseValueException`` to be thrown from it.  This indicates that the
+value cannot be parsed as input into an appropriate form.  You must not allow other runtime exceptions to escape this method to get
+the normal graphql behaviour for validation.  You MUST return a non null value.
+
+* The ``parseLiteral`` MUST NOT allow any exceptions to be thrown from it.  Instead it returns ``null`` if the AST literal value
```
Comment (andimarek) [inline docs/scalars.rst]: I think we copied the exact behaviour of the ref implementation, but reading this again it seems inconsistend: two times exception and one null value. maybe we should explain it more (if we can) or change it.

Comment (bbakerman) [reply docs/scalars.rst]: I am happy to change this to be consistent, that is throw a CoercingParseValueException for parseLiteral.

That way is way more consistent

File: docs/scalars.rst [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+* The ``serialise`` MUST ONLY allow ``graphql.schema.CoercingSerializeException`` to be thrown from it.  This indicates that the
+value cannot be serialised into an appropriate form.  You must not allow other runtime exceptions to escape this method to get
+the normal graphql behaviour for validation.  You MUST return a non null value
+
+
+* The ``parseValue`` MUST ONLY allow ``graphql.schema.CoercingParseValueException`` to be thrown from it.  This indicates that the
+value cannot be parsed as input into an appropriate form.  You must not allow other runtime exceptions to escape this method to get
+the normal graphql behaviour for validation.  You MUST return a non null value.
+
+* The ``parseLiteral``  MUST ONLY allow ``graphql.schema.CoercingParseValueException`` to be thrown from it.  This indicates that the
```
Comment (andimarek) [inline docs/scalars.rst]: I really think we should rename it to `CoercingParseValueOrLiteralException` or something similar: I am happy to use the same exception in order to not make it not too difficult, but `CoercingParseValueException` is a perfect macht for `parseValue` and it is just confusing to use it it for `parseLiteral` too without aligning the name.

**Comment by bbakerman** [comment]
Ok I have updated it to have a new CoercingParseLiteralException - you convinced me on the symettry argument to have a new exception

**Pull request merged**